### ISSUE TITLE
[HOLD for payment 2024-10-17] [$500] Chat - Edit composer is focused even if emoji picker is open

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#27612]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open any report
2. Open edit composer for 2 messages
3. Open emoji picker for first one
4. Open emoji picker of second one without closing previous message emoji picker

## Expected Result:
 Edit Composer should not be focused if emoji picker is displayed

## Actual Result:
 Edit Composer is focused if emoji picker is displayed

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] Windows / Chrome

**Version Number:** 1.3.70-5

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/488001a6-9f6b-4dca-9fea-af34e924bc14

https://github.com/Expensify/App/assets/78819774/592b30c0-8304-4af8-9846-efbf7ea68a69

**Expensify/Expensify Issue URL:**

**Issue reported by:** @krishna2323)

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1694848738372809

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~016c6dc7febe9f77e8</li>
        <li>Upwork Job ID: 1703183624812048384</li>
        <li>Last Price Increase: 2024-09-25</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @johncschuster</details>

**Comment by tanerochris** [comment]
## Proposal

## Please re-state the problem that we are trying to solve in this issue.
Report Edit Message composer remains focused when emoji picker remains open. 

## What is the root cause of that problem?
The issue stems from the fact that closing the emoji  picker sets the isFocused state to true, this remains true when opening the emoji picker of another edit box.
https://github.com/Expensify/App/blob/c77814fd5e643cba78e8c8294370d78f633a59c5/src/pages/home/report/ReportActionItemMessageEdit.js#L411C1-L414C31

## What changes do you think we should make in order to solve the problem?

Set  `isFocusedRef.current` is `false`  if not in active edit edit message box. Also to ensure that after selecting an emoji the current edit box becomes focused, we add `focus(true)` at the end of  the `addEmojiToTextBox` handler.

```onModalHide={() => {
                                if (!isFocusedRef.current) setIsFocused(false) 
                                else {
                                    setIsFocused(true);
                                    focus(true);
                                }
                            }}```
      ``` const addEmojiToTextBox = (emoji) => {
        setSelection((prevSelection) => ({
            start: prevSelection.start + emoji.length + CONST.SPACE_LENGTH,
            end: prevSelection.start + emoji.length + CONST.SPACE_LENGTH,
        }));
        updateDraft(ComposerUtils.insertText(draft, selection, `${emoji} `));
        focus(true)
    };
```

**Comment by tanerochris** [comment]
Contributor details
Your Expensify account email: tanejuth@gmail.com
Upwork Profile Link: [LINK](https://www.upwork.com/freelancers/~01645630d3472337f7)

**Comment by ilkeruyanik** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When emoji picker is closed, text input is focused even if it should not focus.

### What is the root cause of that problem?

EmojiPickerButton Component's onModalHide property sets focus without any condition in ReportActionItemMessageEdit Component, when emoji picker is closed.


### What changes do you think we should make in order to solve the problem?
Instead of setting focus directly, we should use ReportActionComposeFocusManager.
https://github.com/Expensify/App/blob/c77814fd5e643cba78e8c8294370d78f633a59c5/src/pages/home/report/ReportActionItemMessageEdit.js#L411-L414

These lines will be changed with
```
onModalHide={() => {
    InteractionManager.runAfterInteractions(() => {
        if (EmojiPickerAction.isEmojiPickerVisible()) {
            return;
        }
        ReportActionComposeFocusManager.focus();
    });
}}
```
New property 'onPress' will be added to EmojiPickerButton component. We will send a function for run on onPress.
In this case, we will use it for set ReportActionComposeFocusManager.onComposerFocus function.
```
onPress={() => {
    ReportActionComposeFocusManager.onComposerFocus(() => {
        if (!textInputRef.current) {
            return;
        }

        textInputRef.current.focus();
    });
}}
```
Screencast for showing the bug is not just about changing emoji picker

https://github.com/Expensify/App/assets/93402058/8112d40f-6754-4385-bfed-f4bb0519599f



### Result
https://github.com/Expensify/App/assets/93402058/28ecafc6-a029-4c64-99d3-654da1cca983

**Comment by ilkeruyanik** [comment]
Contributor details
Your Expensify account email: ilker.uyanik11@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01c158240bd3a5fc99

**Comment by johncschuster** [comment]
Sounds good! Let's bump this one to `Weekly` while we wait.

**Comment by johncschuster** [comment]
Still holding on https://github.com/Expensify/App/issues/25892

**Comment by johncschuster** [comment]
Still holding

**Comment by johncschuster** [comment]
Still holding

**Comment by johncschuster** [comment]
Still holding

**Comment by johncschuster** [comment]
Still holding

**Comment by johncschuster** [comment]
Still holding

**Comment by johncschuster** [comment]
Still holding

**Comment by johncschuster** [comment]
Still holding

**Comment by johncschuster** [comment]
I will be OOO starting Monday, December 18, and will return Tuesday, January 2.

**Current status:** 
This issue is on hold for #25892

If this issue is open when I'm back from OOO, I'll take it back over. Thank you!

**Comment by abekkala** [comment]
this is still on hold for 25892 - changing this to weekly as daily is not needed while on hold

**Comment by abekkala** [comment]
I was assigned as John's second person while he is ooo - however, I am also ooo. 
Johns status is here: https://github.com/Expensify/App/issues/27612#issuecomment-1858544587

This is on hold for #25892

**Comment by johncschuster** [comment]
Thanks for taking care of this while I was OOO @abekkala and @kevinksullivan. I'll take it back over.

**Comment by johncschuster** [comment]
Still on hold. Bumping to Weekly while we wait.

**Comment by johncschuster** [comment]
Still on hold for https://github.com/Expensify/App/issues/25892

**Comment by johncschuster** [comment]
Still on hold for https://github.com/Expensify/App/issues/25892

**Comment by eVoloshchak** [comment]
Not overdue, still on hold for https://github.com/Expensify/App/issues/25892

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold. Also, I've filed this under #vip-vsb since it's related to chat.

**Comment by eVoloshchak** [comment]
https://github.com/Expensify/App/issues/25892 is still open, on HOLD

**Comment by johncschuster** [comment]
Still on hold

**Comment by eVoloshchak** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by eVoloshchak** [comment]
Not overdue, still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold. It looks like the PR in the linked issue is under review, so we may be able to take this one off hold soon.

**Comment by johncschuster** [comment]
Same as above. Waiting on the PR review for the linked issue

**Comment by johncschuster** [comment]
Same as above

**Comment by eVoloshchak** [comment]
https://github.com/Expensify/App/pull/28238 we were holding on was merged 3 days ago
This didn't resolve the bug fully, the composer is still focused for a moment when emoji picker is opened

https://github.com/user-attachments/assets/6b6d64c0-0946-4755-9a79-ba31516288c9

**Comment by johncschuster** [comment]
Thanks! Taking this off hold so we can get someone to work on this.

**Comment by s77rt** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Composer is briefly focused after opening another emoji picker.

### What is the root cause of that problem?
After closing the (first) emoji picker, the [`onModalHide`](https://github.com/Expensify/App/blob/71db3f19691069182447b6b1b9fff62ec93be86c/src/pages/home/report/ReportActionItemMessageEdit.tsx#L569-L571) callback put the focus on the composer.

### What changes do you think we should make in order to solve the problem?
In the `onModalHide` callback, return early if the current active/focused element is a composer or an emoji picker button.

```js
const activeElement = DomUtils.getActiveElement();
if (activeElement && [messageEditInput, emojiButtonID].includes(activeElement.id)) {
    return;
}
```

### What alternative solutions did you explore? (Optional)
In the `onModalHide` callback, return early if we are going to open a new modal i.e. `Modal.willAlertModalBecomeVisible()`. Unfortunately this approach won't work because `onModalHide` is called way before creating the new modal.

**Comment by eVoloshchak** [comment]
@tienifr's [proposal](https://github.com/Expensify/App/issues/27612#issuecomment-2380318226) looks good to me!

🎀👀🎀 C+ reviewed!

**Comment by s77rt** [comment]
I think I'm supposed to be assigned here, right? @eVoloshchak 
PR: https://github.com/Expensify/App/pull/50099

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: https://github.com/Expensify/App/pull/24482
- The offending PR has been commented on: https://github.com/Expensify/App/pull/24482#discussion_r1801472095
- A discussion in #expensify-bugs has been started: Not needed
- Determine if we should create a regression test for this bug: I don't think this requires a regression test

**Comment by eVoloshchak** [comment]
@johncschuster, should this be $500 or $250? The issue was originally $500 and I don't see any price changes, but that was set a long time ago

**Comment by johncschuster** [comment]
Thanks for asking, @eVoloshchak! Let me chat with the team quickly about that. I'll be back shortly!

**Comment by johncschuster** [comment]
# Payment Summary
Contributor+: @eVoloshchak owed $250 via NewDot
Contributor: @s77rt owed $250 via NewDot

**Comment by JmillsExpensify** [comment]
$500 approved for @eVoloshchak

**Comment by JmillsExpensify** [comment]
$500 approved for @s77rt
